src/yarp.c

```diff
@@ -3355,21 +3377,90 @@ lex_token_type(yp_parser_t *parser) {
         }
       }
 
+      // If we're in a tilde heredoc, we need to calculate the minimum amount of leading
+      // whitespace so that we can "ignore" this amount of whitespace on all lines in the
+      // heredoc.
+      //
+      // In order to calculate the minimum amount of whitespace, we find the end of the
+      // heredoc, and then iterate from the start until the end, counting the whitespace
+      // on every line and using the `ignored_whitespace` field on the heredoc struct to
+      // track the minimum whitespace we see.
+      if (heredoc_struct.indent == YP_HEREDOC_INDENT_TILDE &&
```

Comment:
Originally, I thought this was correct, but looking at it more, I'm not so sure now.

Consider this very annoying example:

```ruby
p <<~FOO
    #{<<~FOO}
        inner
    FOO
  outer
FOO
```

In the snippet above, this block will loop until it finds the first `FOO` and then it will be very confused.

I don't think we can eagerly look forward and just match blindly without running the parser.

The other example doesn't involve nesting, just confusing interpolation:

```ruby
p <<~FOO
    #{
FOO
    }
FOO
```

I think we're going to have to have the context of the parser in order to know what we're allowed to consider as a terminator.

I don't really know the solution here, or how CRuby does it. Maybe it doesn't emit any of the string content events until it has parsed the full heredoc? I'm not sure.